tests/lax_scipy_special_functions_test.py

```diff
@@ -172,13 +176,14 @@ def _GetArgsMaker(self, rng, shapes, dtypes):
       shapes=itertools.combinations_with_replacement(all_shapes, rec.nargs),
       dtypes=(itertools.combinations_with_replacement(rec.dtypes, rec.nargs)
         if isinstance(rec.dtypes, list) else itertools.product(*rec.dtypes)),
+      test_name=[rec.test_name],
     ))
     for rec in JAX_SPECIAL_FUNCTION_RECORDS
   ))
   @jax.numpy_rank_promotion('allow')  # This test explicitly exercises implicit rank promotion.
   @jax.numpy_dtype_promotion('standard')  # This test explicitly exercises dtype promotion
   def testScipySpecialFun(self, op, rng_factory, shapes, dtypes,
-                          test_autodiff, nondiff_argnums):
+                          test_autodiff, nondiff_argnums, test_name):
```

Comment:
Can you say more about why this change is necessary? It doesn't look like `test_name` is used in the function.

Reply:
In my chain of thought, the `test_name` variable is needed for generating the actual test name in the `pretty_special_fun_name` (to distinguish the test name for the new test case). However, `pytest.parameterize` carries whole values to the decorated function, so I just add them even though they are not used in the function body. I couldn't figure out how to do it without adding the new function argument, but if it is, I will change my implementation.

Reply:
Ah I see - it looks like the `test_name` is not used anymore after we refactored these tests last year. I think this is the wrong approach.

Instead of adding more test cases, let's leave the previous test cases alone and add a new manual test for each at the bottom of the file. What do you think?

Reply:
Actually, another thought if you're willing to look into it: I think the semantics of `jax.scipy.special.xlogy` match what we'd need for `rel_entr`, and there's a custom JVP rule to correctly handle gradients near zero. I'd bet if we re-implemented `rel_entr` in terms of `xlogy`, we could make the existing test work with the expanded domain. Is that something you want to explore?

Reply:
(Alternatively, you could just limit your change to the fix: replacing `q` with `zero` and I'll follow-up on the autodiff & testing questions)

Reply:
I will re-implement it with xlogy.

Reply:
It seems like `xlogy` is tested only in the domain of positive values. I've tried to test `xlogy` with `jtl.rand_some_zero`, but it could not pass the test. I think the jvp rule shows some instability. I change the division operator with `lax.div` shows some stable results, but they failed on two test cases, named `LaxScipySpcialFunctionsTest::testScipySpecialFun_xlogy_3x4_3x4_float32_float32` and `LaxScipySpcialFunctionsTest::testScipySpecialFun_xlogy_3x1_2x1x4_float32_float32`.

I already implemented rel_entr with xlogy. So, could you imagine why the jvp rule of `xlogy` is not work on the two testcases? I couldn't figure it out..

Reply:
```
E           AssertionError:
E           Not equal to tolerance rtol=0.3, atol=0.003
E           JVP tangent
E           x and y -inf location mismatch:
E            x: array([[0.],
E                  [0.],
E                  [0.]], dtype=float32)
E            y: array([[  0.],
E                  [-inf],
E                  [  0.]], dtype=float32)
```
for `LaxScipySpcialFunctionsTest.testScipySpecialFun_xlogy_3x1_2x1x4_float32_float32`.

```
E           AssertionError:
E           Not equal to tolerance rtol=1.2, atol=0.012
E           JVP tangent
E           x and y nan location mismatch:
E            x: array([[nan, nan, nan, nan],
E                  [nan, nan, nan, nan],
E                  [nan, nan, nan, nan]], dtype=float32)
E            y: array([[-inf, -inf, -inf, -inf],
E                  [  0.,   0.,   0.,   0.],
E                  [  0.,   0.,   0.,   0.]], dtype=float32)
```
For `LaxScipySpcialFunctionsTest.testScipySpecialFun_xlogy_3x4_3x4_float32_float32`.

Reply:
OK, then I was probably wrong about `xlogy` being applicable here. Thanks for checking.

Reply:
Let's sort out these things.

First, I checked the behavior of `scipy.special.xlogy`.
```
>>> from scipy.special import xlogy
>>> xlogy([[1.1],[0],[-1.1]], [[1.1,0,-1.1]])
array([[ 0.1048412,       -inf,        nan],
       [ 0.       ,  0.       ,  0.       ],
       [-0.1048412,        inf,        nan]])
```
It seems that the output is well defined when `y>=0` or `x==0` (not nan region).

Second, I think whether the gradient, e.g., jvp, tangent, etc, could be computed through numerical approach, (f(x+eps)-f(x-eps))/(2*eps).

I've checked the code for testing the difference between custom jvp and numerical difference, and find out that they does not consider the nan value (only consider inf) and it is natural as the gradient could not be defined with nan value.

https://github.com/google/jax/blob/6fe313c6a1282158dfb0dccce61e65942368667d/jax/_src/public_test_util.py#L211-L215

https://github.com/google/jax/blob/6fe313c6a1282158dfb0dccce61e65942368667d/jax/_src/public_test_util.py#L187-L188

https://github.com/google/jax/blob/6fe313c6a1282158dfb0dccce61e65942368667d/jax/_src/public_test_util.py#L169-L173

In this case, when we compute the jvp tangent at y=0, using numerical approach could not be valid as f(x-eps) would be nan.
 
Third, I think about a general approach to compute the numerical gradient at the boundary case and one possible solution is to use left gradient or right gradient, (f(x)-f(x-eps))/eps or (f(x+eps)-f(x))/eps where the terms in the numerator would be valid in this case.

However, I am scared of changing the `check_grad` in `public_test_util.py` as it might broken lots of existing testcases, so I have to think about other approaches..

Reply:
Right, I think I was wrong that `xlogy` has relevant semantics. It doesn't and you should ignore that suggestion.